bridges/snowbridge/pallets/system-frontend/src/lib.rs

```diff
@@ -164,52 +184,66 @@ pub mod pallet {
 
-			let mut checked = false;
-			if asset_location.eq(&origin_location) || asset_location.starts_with(&origin_location) {
-				checked = true
-			}
-			ensure!(checked, <Error<T>>::OwnerCheck);
+			// Burn Ether Fee for the cost on ethereum
+			Self::burn_for_teleport(&origin_location, &(T::EthereumLocation::get(), fee).into())?;
 
-			Self::burn_fees(origin_location.clone(), fee)?;
+			let reanchored_asset_location = Self::reanchor(&asset_location)?;
 
-			let call = SnowbridgeControl::Control(ControlCall::RegisterToken {
-				asset_id: Box::new(VersionedLocation::from(asset_location.clone())),
+			let call = BridgeHubRuntime::EthereumSystem(EthereumSystemCall::RegisterToken {
+				asset_id: Box::new(VersionedLocation::from(reanchored_asset_location.clone())),
 				metadata,
 				fee,
 			});
 
-			let xcm: Xcm<()> = vec![
-				ReceiveTeleportedAsset(T::DeliveryFee::get().into()),
-				PayFees { asset: T::DeliveryFee::get() },
-				Transact {
-					origin_kind: OriginKind::Xcm,
-					call: call.encode().into(),
-					fallback_max_weight: None,
-				},
-			]
-			.into();
+			let message_id = Self::send(origin_location.clone(), Self::build_xcm(call.encode()))?;
 
-			Self::send(xcm)?;
-
-			Self::deposit_event(Event::<T>::RegisterToken { location: asset_location });
+			Self::deposit_event(Event::<T>::RegisterToken { location: asset_location, message_id });
 
 			Ok(())
 		}
 	}
 
 	impl<T: Config> Pallet<T> {
-		pub fn send(xcm: Xcm<()>) -> DispatchResult {
-			let bridgehub = Location::new(1, [Parachain(1002)]);
-			send_xcm::<T::XcmSender>(bridgehub, xcm).map_err(|_| Error::<T>::Send)?;
-			Ok(())
+		fn send(origin: Location, xcm: Xcm<()>) -> Result<H256, Error<T>> {
+			let (message_id, price) =
+				send_xcm::<T::XcmSender>(T::BridgeHubLocation::get(), xcm.clone()).map_err(
+					|err| {
+						tracing::error!(target: LOG_TARGET, ?err, ?xcm, "XCM send failed with error");
+						Error::<T>::from(err)
+					},
+				)?;
+			T::XcmExecutor::charge_fees(origin, price).map_err(|_| Error::<T>::FundsUnavailable)?;
+			Ok(message_id.into())
 		}
-		pub fn burn_fees(origin_location: Location, fee: u128) -> DispatchResult {
-			let ethereum_fee_asset = (T::FeeAsset::get(), fee).into();
-			T::AssetTransactor::withdraw_asset(&ethereum_fee_asset, &origin_location, None)
+
+		fn burn_for_teleport(origin: &Location, fee: &Asset) -> DispatchResult {
+			let dummy_context =
+				XcmContext { origin: None, message_id: Default::default(), topic: None };
+			T::AssetTransactor::can_check_out(origin, fee, &dummy_context)
 				.map_err(|_| Error::<T>::FundsUnavailable)?;
-			T::AssetTransactor::withdraw_asset(&T::DeliveryFee::get(), &origin_location, None)
+			T::AssetTransactor::check_out(origin, fee, &dummy_context);
+			T::AssetTransactor::withdraw_asset(fee, origin, None)
 				.map_err(|_| Error::<T>::FundsUnavailable)?;
 			Ok(())
 		}
+
+		fn build_xcm(call: Vec<u8>) -> Xcm<()> {
```

Comment:
We shouldn't pass raw bytes around unless required. Consider constraining `call` as follows:

```suggestion
		/// Build an XCM to invoke `call` on the destination parachain
		fn build_xcm(call: &impl Encode) -> Xcm<()> {
```

Or just pass `call: BridgeHubRuntime`

Reply:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/263abab30af616000d82065262078068e878975f